### PR TITLE
Introduce the ability to transform the parsed tree by applying a mapping function to each tree node recursively

### PR DESCRIPTION
**Pull request by izeigerman**

### CODE REVIEW

**Comment by izeigerman** [comment]
The error from linter is:
```
sqlglot/parser.py:529:11: R1714: Consider merging these comparisons with "in" to 'this.token_type not in (TokenType.ALIAS, TokenType.TABLE)' (consider-using-in)
```
Doesn't seem like I was the one who introduced this.

@tobymao I can try and fix this linter error as part of this PR if that's ok.

File: sqlglot/expressions.py [hunk 1]
```diff
@@ -105,6 +105,45 @@ def to_s(self, level=0):
...
+
+        Returns:
+            the transformed tree.
+        """
+        node = deepcopy(self) if copy else self
+        new_node = fun(node)
+
+        if new_node is None:
+            raise ValueError('A transformed node cannot be None')
+        if not isinstance(new_node, Expression) or new_node is not node:
```
Comment (tobymao) [inline sqlglot/expressions.py:127]: new_node != node, what's the differenece with is?

File: sqlglot/expressions.py [hunk 2]
```diff
@@ -105,6 +105,45 @@ def to_s(self, level=0):
...
+
+        for k, v in new_node.args.items():
+            is_list_arg = isinstance(v, list)
+
+            child_nodes = v if is_list_arg else [v]
+            new_child_nodes = []
+
+            for cn in child_nodes:
+                if isinstance(cn, Expression):
+                    new_child_node = cn.transform(fun, copy=False)
```
Comment (tobymao) [inline sqlglot/expressions.py:138]: why is copy False here, and not inherit copy from caller?

File: sqlglot/expressions.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/expressions.py:127]: Later if we decide to introduce a proper equality comparison, I figured this way we won't have to compare entire subtrees. In other words I think this way is cheaper computation-wise. What do you think?

File: sqlglot/expressions.py (hunk 2 above)
Comment (izeigerman) [reply sqlglot/expressions.py:138]: Given that we already performed a deep copy of the top node I see no reason to perform it again for children nodes since they've been copied already. Let me know if I'm missing something here.

File: tests/test_expressions.py [hunk 3]
```diff
@@ -52,3 +52,41 @@ def test_validate(self):
 
         with self.assertRaises(ValueError):
             exp.Hint()
+
+    def test_transform_simple(self):
+        expression = parse('IF(a > 0, a, b)')[0]
```
Comment (tobymao) [inline tests/test_expressions.py:57]: i wonder if we want to add a new api parse_one or something like that or if [0] is fine everywhere

File: sqlglot/expressions.py [hunk 4]
```diff
@@ -12,7 +13,6 @@ def __init__(self, **args):
         self.args = args
         self.validate()
         self._parent = None
-        self.arg_key = None
```
Comment (tobymao) [inline sqlglot/expressions.py:15]: why is this removed?

Comment (izeigerman) [reply sqlglot/expressions.py:15]: Ah, good catch. Accidental change. Btw, I don't see it being used anywhere except when we set it during parser. What's it for?

File: tests/test_expressions.py (hunk 3 above)
Comment (izeigerman) [reply tests/test_expressions.py:57]: I was wondering the same thing. I think having `parse_one` might be a good idea. I personally like it more than [0].

File: sqlglot/expressions.py (hunk 4 above)
Comment (tobymao) [reply sqlglot/expressions.py:15]: it's so when you walk, when you yield, you know what the parent's key of the token was

File: sqlglot/expressions.py (hunk 1 above)
Comment (tobymao) [reply sqlglot/expressions.py:127]: ok sounds good

File: sqlglot/expressions.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/expressions.py:138]: got it, you're right

File: tests/test_expressions.py (hunk 3 above)
Comment (tobymao) [reply tests/test_expressions.py:57]: ok let's add that and refactor

Comment (izeigerman) [reply tests/test_expressions.py:57]: Do you mean add that in this PR? Or a subsequent one? I'm ok with either way

Comment (izeigerman) [reply tests/test_expressions.py:57]: Never mind, I've already refactored this one. Will push the change soon.

**Review by tobymao — APPROVED** [review]
